### DISCUSSION TITLE
How to identify coupling on large S parameter file

### DISCUSSION THREAD

**Discussion by amichel0205** [ansys/pyaedt#7318] [Show and tell]
Analyzing a large S-parameter file (for example, one generated from an RF board with a high number of ports) is extremely difficult—if not impossible—using a conventional rectangular plot. Moreover, applying filters such as frequency range and/or insertion loss limits is essential for meaningful analysis. A PyAEDT function named get_coupling_in_range() was created for this purposed.

In this discussion you will find a script example on how to use function  get_coupling_in_range() . 
A 16 ports  S parameter file, extracted from HFSS simulation, is used as test case.

It is mandatory that the port naming in the HFSS project follow some naming convention: 

ComponentName_PinName e.g.  U1_AG11

We will use a toml file to pass  parameter to the script.

The toml file below has to be place in AEDT personallib.

```` python
[path]
# Path to S parameter file
sparampath = "D:\\tmp\\coupling_test\\si_verse_cutout.s16p"
# Path to log file
logfilename = "D:\\tmp\\coupling_test\\Sparameter_test.log"

[parameter]
# list of parameters to set up coupling identification
#Check of coupling will:
# -start at startatfrequency in Hz
startatfrequency = 2e9
# -stop at stopatfrequency in Hz
stopatfrequency = 4e9
# -sample every freqsample
freqsample = 5
# -identifying S parameter with value in dB between greater than highloss and less than lowloss
highloss = -40
lowloss = -20
# exclude S parameter from pins of same component exclude = false
# e.g. (Scomp1_xx, scomp1_yy) excluded
# only works if component port naming is as follow: 
# comp1_a1 where comp1 is component property Name and a1 pin name to which the port is connected to
ssamecomponent = false
# list of components
# e.g. list = ["comp1", "comp2") to which parameter excludeinclude will apply
# only works if component port naming is as follow: 
# comp1_a1 where comp1 is component property Name and a1 pin name to which the port is connected to
complist = []
# exclude list = false include only list = true
excludeinclude = true
# -to plot s parameter curves set plot to true to not plot s parameter curves set plot to false
plot = true
````
Each parameter usage is explain in the toml file.
The script below will :

- Read toml file located in AEDT personallib
- run PyAEDT function get_coupling_in_range() using parameters set in toml file
- Write log file
- From information of log file plot S parameter identified if parameter plot is set to true

````python
# -*- coding: utf-8 -*-
#
# Copyright (C) 2021 - 2025 ANSYS, Inc. and/or its affiliates.
# SPDX-License-Identifier: MIT
#
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

from ansys.aedt.core.desktop import Desktop
from ansys.aedt.core.aedt_logger import pyaedt_logger as logger
from ansys.aedt.core.visualization.advanced.touchstone_parser import TouchstoneData
from ansys.aedt.core.generic.file_utils import open_file
import os
import toml
import matplotlib.pyplot as plt

# AEDT version
version = "2024.2"

aedttmp = Desktop(version=version, non_graphical=True)
personallib = aedttmp.personallib
print(personallib)
aedttmp.close_desktop()

print("This toml file will setup parameter for identifying coupling: "
      , os.path.join(personallib, "Coupling_identification_setup.toml"))

# Setting of toml configuration file used to set a dictionary
res = toml.load(os.path.join(personallib, "Coupling_identification_setup.toml"))
# Path to S parameter file
s_param_path = res['path']['sparampath']
# Path to log file
log_file_name = res['path']['logfilename']
# Specify frequency value below which not check will be done
start_at_frequency = res['parameter']['startatfrequency']
# Specify frequency value above which not check will be done
stop_at_frequency = res['parameter']['stopatfrequency']
# identifying S parameter with value in dB between greater high_loss and less than low_loss
low_loss = res['parameter']['lowloss']
high_loss = res['parameter']['highloss']
# exclude S parameter from pins of same component exclude = false
s_same_component = res['parameter']['ssamecomponent']
# Component list
comp_list = res['parameter']['complist']
# component from list exclude = False or only include = True
exclude_include = res['parameter']['excludeinclude']
# Specify frequency sample at which coupling check will be done
freq_sample = res['parameter']['freqsample']
plot = res['parameter']['plot']

if os.path.isfile(s_param_path):
    sp = TouchstoneData(touchstone_file=s_param_path)

    temp_list = sp.get_coupling_in_range(start_frequency=start_at_frequency, stop_frequency=stop_at_frequency,
                                         high_loss=high_loss, low_loss=low_loss,
                                         include_same_component=s_same_component,
                                         component_filter=comp_list, include_filter=exclude_include,
                                         frequency_sample=freq_sample, output_file=log_file_name)

    temp_list_port_name = []
    if os.path.exists(log_file_name):
        with open_file(log_file_name, "r") as fr:
            line = fr.readlines()
            fr.close()
        with open_file(log_file_name, "w") as fw:
            for l in line:
                port1 = (l[(l.find("(") + 1):(l.find(",") - 1)])
                port2 = (l[(l.find(",") + 1):(l.find(")"))])
                previous_string = (l[:(l.find("\n") - 1)])
                new_string = f"{previous_string}, {port1}, {port2}\n"
                fw.write(new_string)
            fw.close()

    if plot:
        i = 0
        while i < len(temp_list):
            sp.plot_s_db(temp_list[i][0], temp_list[i][1])
            i = i + 1
        logger.info("Plotting S parameters curves.")
        plt.show()
        logger.info("S parameters curves plot created.")
else:
    logger.error("File " + s_param_path + " does not exist.")

````

Below is the results of the log file created.

<img width="630" height="381" alt="image" src="https://github.com/user-attachments/assets/8c096846-eae7-4471-93d8-3dc042e5eb52" />

If you have any question do not hesitate to use comment window from this discussion
